File: Day-01/typescript/zFlxw/day01.ts
// Made using the Deno.land runtime. The utils.ts file can be found in /shared/zFlxw.

import { readFileLines } from '../../../shared/zFlxw/utils.ts';

export function day01_01() {
  const lines = readFileLines('day01/input.txt');
  let totalSum = 0;
  for (const line of lines) {
    const digits: number[] = line.split('').filter(char => !isNaN(+char)).map(digit => +digit);
    totalSum += (digits[0] * 10) + digits[digits.length - 1];
  }

  console.log('Total sum of all digits: ' + totalSum);
}

export function day01_02() {
  const digits = {
    'one': 1,
    'two': 2,
    'three': 3,
    'four': 4,
    'five': 5,
    'six': 6,
    'seven': 7,
    'eight': 8,
    'nine': 9
  }

  const lines = readFileLines('input.txt');

  let totalSum = 0;
  for (const line of lines) {
    const numbers: number[] = []; 

    for (let i = 0; i < line.length; i++) {
      const char = line.split('')[i];
      if (!isNaN(+char)) {
        numbers.push(+char)
        continue;
      }

      for (const key of Object.keys(digits)) {
        if (line.slice(i).startsWith(key)) {
          numbers.push(+digits[key as keyof typeof digits]);
        }
      }
    }
    
    totalSum += numbers[0] * 10 + numbers[numbers.length - 1];
  }

  console.log("Total Sum: " + totalSum);
}

File: Day-01/elixir/icrayix/day01.exs
lines =
  String.split(File.read!("input.txt"), "\n", trim: true)

# part 1
Enum.map(lines, fn line ->
  # 57 is the unicode codepoint for '9'
  line = String.to_charlist(line) |> Enum.filter(fn char -> char <= 57 end)

  (Enum.take(line, 1) ++ Enum.take(line, -1))
  |> List.to_integer()
end)
|> Enum.sum()
|> IO.puts()

# part 2
defmodule Part2 do
  def parse_nums(""), do: []
  def parse_nums(<<c>> <> rest) when c <= 57, do: [c - ?0, parse_nums(rest)]

  digits =
    Enum.with_index(
      [
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine"
      ],
      1
    )

  for {digit, index} <- digits do
    def parse_nums(<<unquote(digit)>> <> _ = all) do
      <<_>> <> rest = all
      [unquote(index), parse_nums(rest)]
    end
  end

  def parse_nums(<<_>> <> rest), do: [parse_nums(rest)]
end

Enum.map(lines, fn line ->
  line = Part2.parse_nums(line) |> List.flatten()

  Enum.at(line, 0) * 10 + Enum.at(line, -1)
end)
|> Enum.sum()
|> IO.puts()
